lib/Sema/SemaBounds.cpp

```diff
@@ -524,25 +636,25 @@ namespace {
           MemberExpr *M = dyn_cast<MemberExpr>(E);
           if (!M) {
             llvm_unreachable("unexpected cast failure");
-            return CreateBoundsNone();
+            return CreateBoundsInferenceError();
           }
 
           FieldDecl *F = dyn_cast<FieldDecl>(M->getMemberDecl());
           if (!F)
-            return CreateBoundsNone();
+            return CreateBoundsInferenceError();
 
           BoundsExpr *B = F->getBoundsExpr();
           if (!B || B->isNone() || B->isInteropTypeAnnotation())
-            return CreateBoundsNone();
+            return CreateBoundsInferenceError();
 
           Expr *MemberBaseExpr = M->getBase();
           if (!SemaRef.CheckIsNonModifyingExpr(MemberBaseExpr,
               Sema::NonModifiyingExprRequirement::NMER_Unknown,
               /*ReportError=*/false))
-            return CreateBoundsNone();
+            return CreateBoundsNotAllowedYet();
           B = SemaRef.MakeMemberBoundsConcrete(MemberBaseExpr, M->isArrow(), B);
           if (!B)
-            return CreateBoundsNone();
+            return CreateBoundsInferenceError();
```

Comment:
Here it feels like if we failed to make the bounds concrete, then it's our fault, rather than the programmer's. I also just took a look at `MakeMemberBoundsConcrete` and I think it has a logic bug.